### ISSUE TITLE
[Feature] Support for Python 3.12

### ISSUE THREAD

**Issue by cyberw** [microsoft/playwright-python#2096]
Needs an update of `greenlet` to 3.0.0

**Comment by xsoft7** [comment]
i have a greenlet 3.0.0 installed and cannot install playwright 1.38.0.

using latest pycharm 2023.2.2


the error is:

`Collecting playwright
  Obtaining dependency information for playwright from https://files.pythonhosted.org/packages/e1/b0/c59fcdda1a05cdfce282ba5c410d57e696aec614bba9a4f5c43bee7731fc/playwright-1.38.0-py3-none-macosx_11_0_arm64.whl.metadata
  Using cached playwright-1.38.0-py3-none-macosx_11_0_arm64.whl.metadata (3.5 kB)
Collecting greenlet==2.0.2 (from playwright)
  Using cached greenlet-2.0.2.tar.gz (164 kB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
Collecting pyee==9.0.4 (from playwright)
  Using cached pyee-9.0.4-py2.py3-none-any.whl (14 kB)
Requirement already satisfied: typing-extensions in /Users/dannyk/PycharmProjects/pw2/venv/lib/python3.12/site-packages (from pyee==9.0.4->playwright) (4.8.0)
Using cached playwright-1.38.0-py3-none-macosx_11_0_arm64.whl (31.4 MB)
Building wheels for collected packages: greenlet
  Building wheel for greenlet (pyproject.toml): started
  Building wheel for greenlet (pyproject.toml): finished with status 'error'
Failed to build greenlet

  error: subprocess-exited-with-error
  
  × Building wheel for greenlet (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [177 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.macosx-10.9-universal2-cpython-312
      creating build/lib.macosx-10.9-universal2-cpython-312/greenlet
      copying src/greenlet/__init__.py -> build/lib.macosx-10.9-universal2-cpython-312/greenlet
      creating build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/__init__.py -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      creating build/lib.macosx-10.9-universal2-cpython-312/greenlet/tests
      copying src/greenlet/tests/test_version.py -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/tests
      copying src/greenlet/tests/test_weakref.py -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/tests
      copying src/greenlet/tests/test_gc.py -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/tests
      copying src/greenlet/tests/leakcheck.py -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/tests
      copying src/greenlet/tests/test_generator.py -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/tests
      copying src/greenlet/tests/test_greenlet_trash.py -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/tests
      copying src/greenlet/tests/test_throw.py -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/tests
      copying src/greenlet/tests/test_tracing.py -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/tests
      copying src/greenlet/tests/test_cpp.py -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/tests
      copying src/greenlet/tests/test_contextvars.py -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/tests
      copying src/greenlet/tests/test_greenlet.py -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/tests
      copying src/greenlet/tests/test_extension_interface.py -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/tests
      copying src/greenlet/tests/__init__.py -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/tests
      copying src/greenlet/tests/test_generator_nested.py -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/tests
      copying src/greenlet/tests/test_stack_saved.py -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/tests
      copying src/greenlet/tests/test_leaks.py -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/tests
      running egg_info
      writing src/greenlet.egg-info/PKG-INFO
      writing dependency_links to src/greenlet.egg-info/dependency_links.txt
      writing requirements to src/greenlet.egg-info/requires.txt
      writing top-level names to src/greenlet.egg-info/top_level.txt
      reading manifest file 'src/greenlet.egg-info/SOURCES.txt'
      reading manifest template 'MANIFEST.in'
      warning: no previously-included files found matching 'benchmarks/*.json'
      no previously-included directories found matching 'docs/_build'
      warning: no files found matching '*.py' under directory 'appveyor'
      warning: no previously-included files matching '*.pyc' found anywhere in distribution
      warning: no previously-included files matching '*.pyd' found anywhere in distribution
      warning: no previously-included files matching '*.so' found anywhere in distribution
      warning: no previously-included files matching '.coverage' found anywhere in distribution
      adding license file 'LICENSE'
      adding license file 'LICENSE.PSF'
      adding license file 'AUTHORS'
      writing manifest file 'src/greenlet.egg-info/SOURCES.txt'
      copying src/greenlet/greenlet.cpp -> build/lib.macosx-10.9-universal2-cpython-312/greenlet
      copying src/greenlet/greenlet.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet
      copying src/greenlet/greenlet_allocator.hpp -> build/lib.macosx-10.9-universal2-cpython-312/greenlet
      copying src/greenlet/greenlet_compiler_compat.hpp -> build/lib.macosx-10.9-universal2-cpython-312/greenlet
      copying src/greenlet/greenlet_cpython_compat.hpp -> build/lib.macosx-10.9-universal2-cpython-312/greenlet
      copying src/greenlet/greenlet_exceptions.hpp -> build/lib.macosx-10.9-universal2-cpython-312/greenlet
      copying src/greenlet/greenlet_greenlet.hpp -> build/lib.macosx-10.9-universal2-cpython-312/greenlet
      copying src/greenlet/greenlet_internal.hpp -> build/lib.macosx-10.9-universal2-cpython-312/greenlet
      copying src/greenlet/greenlet_refs.hpp -> build/lib.macosx-10.9-universal2-cpython-312/greenlet
      copying src/greenlet/greenlet_slp_switch.hpp -> build/lib.macosx-10.9-universal2-cpython-312/greenlet
      copying src/greenlet/greenlet_thread_state.hpp -> build/lib.macosx-10.9-universal2-cpython-312/greenlet
      copying src/greenlet/greenlet_thread_state_dict_cleanup.hpp -> build/lib.macosx-10.9-universal2-cpython-312/greenlet
      copying src/greenlet/greenlet_thread_support.hpp -> build/lib.macosx-10.9-universal2-cpython-312/greenlet
      copying src/greenlet/slp_platformselect.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet
      copying src/greenlet/platform/setup_switch_x64_masm.cmd -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_aarch64_gcc.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_alpha_unix.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_amd64_unix.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_arm32_gcc.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_arm32_ios.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_arm64_masm.asm -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_arm64_masm.obj -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_arm64_msvc.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_csky_gcc.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_m68k_gcc.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_mips_unix.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_ppc64_aix.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_ppc64_linux.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_ppc_aix.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_ppc_linux.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_ppc_macosx.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_ppc_unix.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_riscv_unix.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_s390_unix.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_sparc_sun_gcc.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_x32_unix.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_x64_masm.asm -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_x64_masm.obj -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_x64_msvc.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_x86_msvc.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/platform/switch_x86_unix.h -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/platform
      copying src/greenlet/tests/_test_extension.c -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/tests
      copying src/greenlet/tests/_test_extension_cpp.cpp -> build/lib.macosx-10.9-universal2-cpython-312/greenlet/tests
      running build_ext
      building 'greenlet._greenlet' extension
      creating build/temp.macosx-10.9-universal2-cpython-312
      creating build/temp.macosx-10.9-universal2-cpython-312/src
      creating build/temp.macosx-10.9-universal2-cpython-312/src/greenlet
      clang -fno-strict-overflow -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -O3 -Wall -arch arm64 -arch x86_64 -g -I/Users/dannyk/PycharmProjects/pw2/venv/include -I/Library/Frameworks/Python.framework/Versions/3.12/include/python3.12 -c src/greenlet/greenlet.cpp -o build/temp.macosx-10.9-universal2-cpython-312/src/greenlet/greenlet.o --std=gnu++11
      In file included from src/greenlet/greenlet.cpp:19:
      In file included from src/greenlet/greenlet_internal.hpp:20:
      src/greenlet/greenlet_greenlet.hpp:831:41: error: no member named 'use_tracing' in '_PyCFrame'
          this->use_tracing = tstate->cframe->use_tracing;
                              ~~~~~~~~~~~~~~  ^
      src/greenlet/greenlet_greenlet.hpp:834:37: error: no member named 'recursion_limit' in '_ts'; did you mean 'py_recursion_limit'?
          this->recursion_depth = tstate->recursion_limit - tstate->recursion_remaining;
                                          ^~~~~~~~~~~~~~~
                                          py_recursion_limit
      /Library/Frameworks/Python.framework/Versions/3.12/include/python3.12/cpython/pystate.h:148:9: note: 'py_recursion_limit' declared here
          int py_recursion_limit;
              ^
      In file included from src/greenlet/greenlet.cpp:19:
      In file included from src/greenlet/greenlet_internal.hpp:20:
      src/greenlet/greenlet_greenlet.hpp:834:63: error: no member named 'recursion_remaining' in '_ts'; did you mean 'c_recursion_remaining'?
          this->recursion_depth = tstate->recursion_limit - tstate->recursion_remaining;
                                                                    ^~~~~~~~~~~~~~~~~~~
                                                                    c_recursion_remaining
      /Library/Frameworks/Python.framework/Versions/3.12/include/python3.12/cpython/pystate.h:150:9: note: 'c_recursion_remaining' declared here
          int c_recursion_remaining;
              ^
      In file included from src/greenlet/greenlet.cpp:19:
      In file included from src/greenlet/greenlet_internal.hpp:20:
      src/greenlet/greenlet_greenlet.hpp:848:42: error: no member named 'trash_delete_nesting' in '_ts'
          this->trash_delete_nesting = tstate->trash_delete_nesting;
                                       ~~~~~~  ^
      src/greenlet/greenlet_greenlet.hpp:867:21: error: no member named 'use_tracing' in '_PyCFrame'
          tstate->cframe->use_tracing = this->use_tracing;
          ~~~~~~~~~~~~~~  ^
      src/greenlet/greenlet_greenlet.hpp:870:13: error: no member named 'recursion_remaining' in '_ts'; did you mean 'c_recursion_remaining'?
          tstate->recursion_remaining = tstate->recursion_limit - this->recursion_depth;
                  ^~~~~~~~~~~~~~~~~~~
                  c_recursion_remaining
      /Library/Frameworks/Python.framework/Versions/3.12/include/python3.12/cpython/pystate.h:150:9: note: 'c_recursion_remaining' declared here
          int c_recursion_remaining;
              ^
      In file included from src/greenlet/greenlet.cpp:19:
      In file included from src/greenlet/greenlet_internal.hpp:20:
      src/greenlet/greenlet_greenlet.hpp:870:43: error: no member named 'recursion_limit' in '_ts'; did you mean 'py_recursion_limit'?
          tstate->recursion_remaining = tstate->recursion_limit - this->recursion_depth;
                                                ^~~~~~~~~~~~~~~
                                                py_recursion_limit
      /Library/Frameworks/Python.framework/Versions/3.12/include/python3.12/cpython/pystate.h:148:9: note: 'py_recursion_limit' declared here
          int py_recursion_limit;
              ^
      In file included from src/greenlet/greenlet.cpp:19:
      In file included from src/greenlet/greenlet_internal.hpp:20:
      src/greenlet/greenlet_greenlet.hpp:881:13: error: no member named 'trash_delete_nesting' in '_ts'
          tstate->trash_delete_nesting = this->trash_delete_nesting;
          ~~~~~~  ^
      src/greenlet/greenlet_greenlet.hpp:891:48: error: no member named 'use_tracing' in '_PyCFrame'
          this->use_tracing = origin_tstate->cframe->use_tracing;
                              ~~~~~~~~~~~~~~~~~~~~~  ^
      src/greenlet/greenlet_greenlet.hpp:899:37: error: no member named 'recursion_limit' in '_ts'; did you mean 'py_recursion_limit'?
          this->recursion_depth = tstate->recursion_limit - tstate->recursion_remaining;
                                          ^~~~~~~~~~~~~~~
                                          py_recursion_limit
      /Library/Frameworks/Python.framework/Versions/3.12/include/python3.12/cpython/pystate.h:148:9: note: 'py_recursion_limit' declared here
          int py_recursion_limit;
              ^
      In file included from src/greenlet/greenlet.cpp:19:
      In file included from src/greenlet/greenlet_internal.hpp:20:
      src/greenlet/greenlet_greenlet.hpp:899:63: error: no member named 'recursion_remaining' in '_ts'; did you mean 'c_recursion_remaining'?
          this->recursion_depth = tstate->recursion_limit - tstate->recursion_remaining;
                                                                    ^~~~~~~~~~~~~~~~~~~
                                                                    c_recursion_remaining
      /Library/Frameworks/Python.framework/Versions/3.12/include/python3.12/cpython/pystate.h:150:9: note: 'c_recursion_remaining' declared here
          int c_recursion_remaining;
              ^
      In file included from src/greenlet/greenlet.cpp:21:
      In file included from src/greenlet/greenlet_slp_switch.hpp:82:
      In file included from src/greenlet/slp_platformselect.h:53:
      src/greenlet/platform/switch_aarch64_gcc.h:71:51: warning: value size does not match register size specified by the constraint and modifier [-Wasm-operand-widths]
                 __asm__ volatile ("mov %0, #0" : "=r" (err));
                                                        ^
      src/greenlet/platform/switch_aarch64_gcc.h:71:35: note: use constraint modifier "w"
                 __asm__ volatile ("mov %0, #0" : "=r" (err));
                                        ^~
                                        %w0
      src/greenlet/greenlet.cpp:3095:36: error: no member named 'trash_delete_nesting' in '_ts'
          return PyLong_FromLong(tstate->trash_delete_nesting);
                                 ~~~~~~  ^
      1 warning and 12 errors generated.
      error: command '/usr/bin/clang' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for greenlet
ERROR: Could not build wheels for greenlet, which is required to install pyproject.toml-based projects
`

**Comment by mxschmitt** [comment]
Blocked by https://github.com/conda-forge/greenlet-feedstock/pull/35

**Comment by mathew-jithin** [comment]
Referring the above, is there any workaround to handle this for now in the case of a Lambda deployment, in linux OS's?

**Comment by songsg149** [comment]
I have same issue on python 3.12
greenlet==3.0.0

Please help me how can I install playwright.

**Comment by mxschmitt** [comment]
Blocked by https://github.com/jfhbrook/pyee/issues/134

**Comment by amgmerce** [comment]
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
playwright 1.41.2 requires greenlet==3.0.3, but you have greenlet 3.0.1 which is incompatible.

I have this error can anyone please help?

**Comment by JuanmaMenendez** [comment]
any update on this ?

**Comment by cyberw** [comment]
You might want to make a new issue. The original problem was resolved long ago.
